fieldservice_recurring/models/fsm_recurring.py

```diff
@@ -0,0 +1,239 @@
+            ['fsm_recurring_id'], ['fsm_recurring_id'])
+        count_data = dict((item['fsm_recurring_id'][0],
+                           item['fsm_recurring_id_count']) for item in data)
+        for recurring in self:
+            recurring.fsm_order_count = count_data.get(recurring.id, 0)
+
+    @api.onchange('fsm_recurring_template_id')
+    def onchange_recurring_template_id(self):
+        if not self.fsm_recurring_template_id:
+            return
+        values = self.populate_from_template()
+        self.update(values)
+
+    def populate_from_template(self, template=False):
+        if not template:
+            template = self.fsm_recurring_template_id
+        vals = {
+            'fsm_frequency_set_id': template.fsm_frequency_set_id,
+            'max_orders': template.max_orders,
+            'description': template.description,
+            'fsm_order_template_id': template.fsm_order_template_id,
+            'company_id': template.company_id,
+        }
+        return vals
+
+    @api.model
+    def create(self, vals):
+        if vals.get('name', _('New')) == _('New'):
+            vals['name'] = self.env['ir.sequence'].next_by_code(
+                'fsm.recurring') or _('New')
+        return super(FSMRecurringOrder, self).create(vals)
+
+    @api.multi
+    def action_start(self):
+        for rec in self:
+            if not rec.start_date:
+                rec.start_date = datetime.now()
+            start_date = fields.Datetime.from_string(rec.start_date)
+            rec._create_order(date=start_date)
+            rec.write({'state': 'progress'})
+
+    @api.multi
+    def action_renew(self):
+        return self.action_start()
+
+    @api.multi
+    def action_cancel(self):
+        for order in self.fsm_order_ids.filtered(
+            lambda o: o.stage_id.is_closed is False
+        ):
+            order.action_cancel()
+        return self.write({'state': 'cancel'})
+
+    def _get_rruleset(self):
+        self.ensure_one()
+        ruleset = rruleset()
+        if self.state != 'progress':
+            return ruleset
+        # set next_date which is used as the rrule 'dtstart' parameter
+        next_date = datetime.now()
+        last_order = self.env['fsm.order'].search([
+            ('fsm_recurring_id', '=', self.id),
+            ('stage_id', '!=', self.env.ref(
+                'fieldservice.fsm_stage_cancelled').id)
+        ], offset=0, limit=1, order='scheduled_date_start desc')
+        if last_order:
+            next_date = fields.Datetime.from_string(
+                last_order.scheduled_date_start)
+        # set thru_date to use as rrule 'until' parameter
+        days_ahead = self.fsm_frequency_set_id.schedule_days
+        request_thru_date = datetime.now() + relativedelta(days=+days_ahead)
+        request_thru_str = request_thru_date.strftime(
+            DEFAULT_SERVER_DATETIME_FORMAT)
+        if self.end_date and (self.end_date < request_thru_str):
```

Comment:
bug when executing cron.
TypeError: unorderable types: datetime.datetime() < str()

Reply:
Fixed, to become like this
        if self.end_date and (self.end_date < request_thru_date):
            thru_date = self.end_date   # Here too, remove the from_string() function
        else:
            thru_date = request_thru_date